rbac/management/group/relation_api_dual_write_group_handler.py

```diff
@@ -123,39 +130,55 @@ def _replicate(self):
         except Exception as e:
             raise DualWriteException(e)
 
-    def replicate_added_role(self, role: Role):
-        """Replicate added role."""
+    def replicate_added_roles(self, roles: Iterable[Role], remove_default_access_from: Optional[TenantMapping] = None):
+        """Generate group relations and binding mapping for role."""
         if not self.replication_enabled():
             return
 
         def add_group_to_binding(mapping: BindingMapping):
             self.group_relations_to_add.append(mapping.add_group_to_bindings(str(self.group.uuid)))
 
-        def create_default_mapping():
-            assert role.system is True, "Expected system role. Mappings for custom roles must already be created."
-            binding = V2rolebinding(
-                str(uuid4()),
-                # Assumes same role UUID for V2 system role equivalent.
-                V2role.for_system_role(str(role.uuid)),
-                V2boundresource(("rbac", "workspace"), str(self.default_workspace.id)),
-                groups=frozenset([str(self.group.uuid)]),
+        for role in roles:
+
+            def create_default_mapping():
```

Comment:
Can we move this function out of the loop? Each iteration of the loop might create a new function object

Reply:
good catch! sure It is possible 👍